werkzeug/routing.py

```diff
@@ -1742,10 +1748,14 @@ def build(self, endpoint, values=None, method=None, force_external=False,
         self.map.update()
         if values:
             if isinstance(values, MultiDict):
-                valueiter = iteritems(values, multi=True)
+                temp = dict(
+                    (k, (values.getlist(k)
+                         if len(values.getlist(k)) > 1
```

Comment:
What about the following? It is imho a tad uglier, but should be pretty optimal in performance (in that there is nothing done which could be spared and keep the functionality the same).

    temp = {}
    for key, list_value in values.lists():
          if len(list_value) == 1:
              temp[key] = list_value[0]
          else:
                 temp[key] = list_value